walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,65 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<Event> eventsAL = new ArrayList<Event>(events);
+      Collection<String> people = request.getAttendees();
+
+      Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+      availableTimes.add(TimeRange.WHOLE_DAY);
+      getAvailableTimes(eventsAL, people, availableTimes, request.getDuration());
+    
+      Collection<TimeRange> availableTimesOptional = new ArrayList<TimeRange>(availableTimes);
+      Collection<String> optionalPeople = request.getOptionalAttendees();
+      getAvailableTimes(eventsAL, optionalPeople, availableTimesOptional, request.getDuration());
+
+      if(request.getAttendees().isEmpty() && availableTimesOptional.isEmpty()){
+          return Arrays.asList();
+      }
+
+      return availableTimesOptional.isEmpty() ? availableTimes : availableTimesOptional;
+  }
+
+  private void getAvailableTimes(Collection<Event> events, Collection<String> people, Collection<TimeRange> availableTimes, long duration){
+      for(Event event : new ArrayList<Event>(events)){
+          if(!Collections.disjoint(event.getAttendees(), people)){
+            createAvailableRange(availableTimes, event.getWhen());
+            events.remove(event);
+          }
+      }
```

Comment:
unsafe behavior to  remove from list while iterating over list. 
You can either remove using an iterator or gather all the things to be removed and removed them at the end

Reply:
changed to an iterator